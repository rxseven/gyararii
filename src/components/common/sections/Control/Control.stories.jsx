import React from 'react';
import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react';

import Control from './index';

const title = 'Components/Common/Sections/Control';
const props = {
  images: ['image-001', 'image-002'],
  onDeleteRequest: action('onDeleteRequest'),
  onDeselect: action('onDeselect'),
  onError: action('onError'),
  onSelectAll: action('onSelectAll')
};

storiesOf(title, module)
  .add('with images', () => <Control {...props} />)
  .add('with selected', () => <Control {...props} selected={['image-001']} />)
  .add('with errors', () => <Control {...props} error={['error-001']} />)
  .add('with status', () => <Control {...props} status="Loading images..." />)
  .add(
    'loading',
    () => <Control {...props} selected={['image-001']} isLoading />,
    {
      notes: 'Disable buttons while a network request is being submitted.'
    }
  );
