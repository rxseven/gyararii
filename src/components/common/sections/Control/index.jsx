import PropTypes from 'prop-types';
import exact from 'prop-types-exact';
import React from 'react';
import { isBrowser, isMobile } from 'react-device-detect';
import styled from 'styled-components';

import Button from 'components/common/base/Button';
import Icon from 'components/common/base/Icon';
import { pluralWord } from 'helpers/string';

const Check = styled(Icon)`
  font-size: 0.75rem;
  margin-right: 4px;
`;

const Count = styled.span`
  color: ${({ theme }) => theme.color.primary};
`;

const Danger = styled.span`
  color: ${({ theme }) => theme.color.danger};
  cursor: pointer;
`;

const Frame = styled.div`
  align-items: center;
  display: flex;
  justify-content: space-between;
`;

const Highlight = styled.strong`
  color: #ff620c;
`;

const Meta = styled.div`
  color: #666;
  font-size: 0.75rem;
  margin-bottom: 0;
`;

const Panel = styled.div`
  align-items: center;
  display: flex;
  justify-content: space-between;
`;

const Summary = styled.span`
  font-size: 0.75rem;
  margin-right: 0.5rem;
`;

const propTypes = exact({
  error: PropTypes.array,
  images: PropTypes.array.isRequired,
  isLoading: PropTypes.bool,
  onDeleteRequest: PropTypes.func.isRequired,
  onDeselect: PropTypes.func.isRequired,
  onError: PropTypes.func.isRequired,
  onSelectAll: PropTypes.func.isRequired,
  selected: PropTypes.array,
  status: PropTypes.string
});

const defaultProps = {
  error: [],
  isLoading: false,
  selected: [],
  status: ''
};

function Control({
  error,
  images,
  isLoading,
  onDeleteRequest,
  onDeselect,
  onError,
  onSelectAll,
  selected,
  status
}) {
  const errorLength = error.length;
  const imageLength = images.length;
  const isEmpty = imageLength < 1;
  const isError = errorLength > 0;
  const selectedLength = selected.length;

  return (
    <Frame>
      <Choose>
        <When condition={selectedLength}>
          <Panel>
            <Summary>
              <If condition={isMobile}>
                <Check icon={['far', 'check-circle']} />
              </If>
              <If condition={imageLength === selectedLength}>All </If>
              <Highlight>{selectedLength}</Highlight>
              <If condition={isBrowser}> selected</If>
            </Summary>
            <Button.Set>
              <Button isLoading={isLoading} onClick={onSelectAll}>
                Select all
              </Button>
              <Button isLoading={isLoading} onClick={onDeselect}>
                Clear
              </Button>
              <Button
                icon="trash-alt"
                isLoading={isLoading}
                minimal
                onClick={onDeleteRequest}
              >
                Delete
              </Button>
            </Button.Set>
          </Panel>
        </When>
        <When condition={!!status}>
          <Meta>{status}</Meta>
        </When>
        <When condition={!isEmpty && !status}>
          <Meta>
            Showing <Count>{imageLength}</Count> images
            {isError && (
              <span>
                {' '}
                with{' '}
                <Danger onClick={onError}>{`${errorLength} ${pluralWord(
                  errorLength,
                  'error'
                )}`}</Danger>
              </span>
            )}
          </Meta>
        </When>
      </Choose>
    </Frame>
  );
}

Control.propTypes = propTypes;
Control.defaultProps = defaultProps;

export default Control;
