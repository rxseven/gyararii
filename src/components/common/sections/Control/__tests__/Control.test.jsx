import { shallow } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';

import Environment from 'tests/environment';
import Control from '../index';

describe('<Control />', () => {
  // Arrange
  const props = {
    images: ['image-001', 'image-002'],
    onDeleteRequest: jest.fn(),
    onDeselect: jest.fn(),
    onError: jest.fn(),
    onSelectAll: jest.fn()
  };
  const component = (
    <Environment>
      <Control {...props} />
    </Environment>
  );

  describe('Unit tests', () => {
    it('should render without crashing', () => {
      // Act
      const wrapper = shallow(component);

      // Assert
      expect(wrapper).toBeDefined();
    });
  });

  describe('Snapshot tests', () => {
    it('should render correctly', () => {
      // Act
      const tree = renderer.create(component).toJSON();

      // Assert
      expect(tree).toMatchSnapshot();
    });
  });
});
