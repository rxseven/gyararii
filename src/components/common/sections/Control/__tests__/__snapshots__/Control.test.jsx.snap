// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Control /> Snapshot tests should render correctly 1`] = `
<div
  className="sc-gzVnrw liWgEh"
>
  <div
    className="sc-dnqmqq guwrEu"
  >
    Showing 
    <span
      className="sc-EHOje lpjwMY"
    >
      2
    </span>
     images
  </div>
</div>
`;
